### PR TITLE
Format: show backtrace when formatting bug found

### PR DESCRIPTION
**Pull request by MakeNowJust**
backtrace is useful to fix a bug.

### CODE REVIEW

**Comment by akzhan** [comment]
It's very useful to maintain autogenerated code.

LGTM.

File: src/compiler/crystal/command/format.cr [hunk 1]
```diff
@@ -182,15 +178,23 @@ class Crystal::Command
...
 
-  private def couldnt_format(file)
+  private def couldnt_format(file, ex)
     STDERR << "Error:".colorize(:red).toggle(@color) << ", " <<
       "couldn't format " << file << ", please report a bug including the contents of it: https://github.com/crystal-lang/crystal/issues"
+    STDERR.puts
+    STDERR.puts
+
+    STDERR.puts ex
+    ex.backtrace.each do |frame|
```
Comment (bew) [inline src/compiler/crystal/command/format.cr]: I thing you shoud use `ex.inspect_with_backtrace(STDOUT)` instead

Comment (Sija) [inline src/compiler/crystal/command/format.cr]: @bew `STDOUT` -> `STDERR`

Comment (bew) [inline src/compiler/crystal/command/format.cr]: true 💤

Comment (MakeNowJust) [inline src/compiler/crystal/command/format.cr]: Thank you!

**Comment by MakeNowJust** [comment]
ping? I really want to merge it.

File: src/compiler/crystal/command/format.cr [hunk 2]
```diff
@@ -68,7 +68,7 @@ class Crystal::Command
           when .invalid_byte_sequence?
             error "'#{result.filename}' is not a valid Crystal source file", exit_code: nil
           when .bug?
-            error "there's a bug formatting '#{result.filename}', please report it including the contents of the file: https://github.com/crystal-lang/crystal/issues", exit_code: nil
+            error "there's a bug formatting '#{result.filename}', to show more information, please run:\n\n  $ crystal tool format '#{result.filename}'", exit_code: nil
```
Comment (RX14) [inline src/compiler/crystal/command/format.cr]: Why not show the backtrace for the file here? Why do you tell the user to run format on the individual file?

File: src/compiler/crystal/command/format.cr [hunk 3]
```diff
@@ -182,15 +178,20 @@ class Crystal::Command
...
-        STDERR.flush
+        couldnt_format "'#{filename}'", ex
       end
     end
   end
 
-  private def couldnt_format(file)
+  private def couldnt_format(file, ex)
     STDERR << "Error:".colorize(:red).toggle(@color) << ", " <<
       "couldn't format " << file << ", please report a bug including the contents of it: https://github.com/crystal-lang/crystal/issues"
```
Comment (RX14) [inline src/compiler/crystal/command/format.cr]: how about

>the contents of the file and this backtrace

?

File: src/compiler/crystal/command/format.cr (hunk 2 above)
Comment (MakeNowJust) [inline src/compiler/crystal/command/format.cr]: - `FormatResult` does not keep an error object currently. I prefer small fix.
  - It is hard to read multiple backtrace, and each of errors is not related to others. I expect to report formatter error with this backtrace.

File: src/compiler/crystal/command/format.cr (hunk 3 above)
Comment (MakeNowJust) [inline src/compiler/crystal/command/format.cr]: me too...

File: src/compiler/crystal/command/format.cr (hunk 2 above)
Comment (RX14) [inline src/compiler/crystal/command/format.cr]: For the multiple-backtrace situation i would expect the formatter to report the first error the same as the single-file mode and then exit.

File: src/compiler/crystal/command/format.cr (hunk 3 above)
Comment (MakeNowJust) [inline src/compiler/crystal/command/format.cr]: Reverted for now.

**Comment by MakeNowJust** [comment]
Resolved conflicts.

File: src/compiler/crystal/command/format.cr (hunk 2 above)
Comment (MakeNowJust) [inline src/compiler/crystal/command/format.cr]: > For the multiple-backtrace situation i would expect the formatter to report the first error the same as the single-file mode and then exit.

This behavior has a big problem: it makes hard to format all files when formatter bug is found. It should continue to format file as possible.

**Comment by MakeNowJust** [comment]
ping. I desire to merge this.

File: src/compiler/crystal/command/format.cr [hunk 4]
```diff
@@ -185,8 +185,19 @@ class Crystal::Command
     end
   end
 
-  private def couldnt_format(file)
-    STDERR << "Error:".colorize(:red).toggle(@color) << ", " <<
-      "couldn't format " << file << ", please report a bug including the contents of it: https://github.com/crystal-lang/crystal/issues"
+  private def couldnt_format(file, ex = nil)
+    STDERR << "Error:".colorize(:red).toggle(@color) << ", "
```
Comment (RX14) [inline src/compiler/crystal/command/format.cr]: Why is there a colon and a comma here? I think that's wrong.

File: src/compiler/crystal/command/format.cr (hunk 2 above)
Comment (RX14) [reply src/compiler/crystal/command/format.cr]: Why not just print all the backtraces then? Justg a simple
```
Error: couldn't format foo.cr, please report a bug including the contents of it
<inspect_with_backtrace>

Error: couldn't format bar.cr, please report a bug including the contents of it
<inspect_with_backtrace>
```

Comment (MakeNowJust) [reply src/compiler/crystal/command/format.cr]: I explained this already: "It is hard to read multiple backtrace", and I worry to open a new issue with unrelated multiple backtraces.

Comment (RX14) [reply src/compiler/crystal/command/format.cr]: And how does making them run the formatter for each file which is broken help that?

Comment (MakeNowJust) [reply src/compiler/crystal/command/format.cr]: They will see new message "please report a bug including the contents of it" by running formatter for each file, it helps them, and I think my solution is kludge enough.

Comment (RX14) [reply src/compiler/crystal/command/format.cr]: I think we need another opinion on this.

**Comment by MakeNowJust** [comment]
Please merge!

**Comment by asterite** [comment]
@MakeNowJust Could you rebase against master? CI is failing in Circle.

**Comment by MakeNowJust** [comment]
rebased

**Pull request merged**